File: 백준/Gold/15661. 링크와 스타트/링크와 스타트.py
def make_diff():
    global diff
    start, link = 0, 0
    for i in range(n):
        for j in range(n):
            if visited[i]==0 and visited[j]==0:
                link += arr[i][j]
            elif visited[i]==1 and visited[j]==1:
                start += arr[i][j]
    diff = min(diff, abs(start-link))
    return

def dfs(level):
    if level==n:
        make_diff()
        return
    visited[level] = 1
    dfs(level+1)
    visited[level] = 0
    dfs(level+1)

n = int(input())
arr = [list(map(int, input().split())) for _ in range(n)]
visited = [0] * n
diff = 1e9

dfs(0)
print(diff)

File: 백준/Gold/15661. 링크와 스타트/링크와 스타트.java
import java.util.*;
import java.io.*;

public class Main {
    static int n;
    static int[][] arr;
    static boolean[] visited;
    static int answer = Integer.MAX_VALUE;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        arr = new int[n][n];
        visited = new boolean[n];

        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        dfs(0, 0);
        System.out.println(answer);
    }

    public static void dfs(int index, int level) {
        if (level > 0 && level < n) {
            int linkScore = 0;
            int startScore = 0;

            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    if (visited[i] && visited[j]) {
                        linkScore += arr[i][j];
                    } else if (!visited[i] && !visited[j]) {
                        startScore += arr[i][j];
                    }
                }
            }

            answer = Math.min(answer, Math.abs(linkScore - startScore));
        }

        for (int i = index; i < n; i++) {
            visited[i] = true;
            dfs(i + 1, level + 1);
            visited[i] = false; 
        }
    }
}